pkg/cmd/cli/cmd/set/env.go

```diff
@@ -139,6 +151,68 @@ func RunEnv(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra.Comman
 		return err
 	}
 
+	var sources []string
+	if secretName != "" {
+		sources = append(sources, fmt.Sprintf("secret/%s", secretName))
+	}
+	if configMapName != "" {
+		sources = append(sources, fmt.Sprintf("configmap/%s", configMapName))
+	}
+
+	if len(sources) != 0 {
+		mapper, typer := f.Object(false)
+		b := resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), kapi.Codecs.UniversalDecoder()).
+			ContinueOnError().
+			NamespaceParam(cmdNamespace).DefaultNamespace().
+			FilenameParam(explicit, false, filenames...).
+			SelectorParam(selector).
+			ResourceTypeOrNameArgs(all, sources...).
+			Flatten()
+
+		one := false
+		infos, err := b.Do().IntoSingular(&one).Infos()
+		if err != nil {
+			return err
+		}
+
+		for _, info := range infos {
+			switch from := info.Object.(type) {
+			case *kapi.Secret:
+				for key := range from.Data {
+					envVar := kapi.EnvVar{
+						Name: fmt.Sprintf("%s%s", prefix, strings.ToUpper(key)),
+						ValueFrom: &kapi.EnvVarSource{
```

Comment:
I know this isn't adding the power, but I don't remember seeing the SA admission plugin getting updated to restrict access to secrets based on the mountable list.

Similarly, how are you protecting configmaps from users who don't have rights to use them?  Not everyone who can create pods will necessarily be able to see other API resources.  I'd be wary of making "create pods" equivalent to "list secrets" and "list configmaps".